cache/src/test/java/io/envoyproxy/controlplane/cache/SimpleCacheTest.java

```diff
@@ -159,6 +184,67 @@ public void successfullyWatchAllResourceTypesWithSetBeforeWatchWithRequestVersio
     }
   }
 
+  @Test
+  public void successfullyWatchAllResourceTypesWithSetBeforeWatchWithSameRequestnVersionNewResourceHints() {
+    SimpleCache<String> cache = new SimpleCache<>(new SingleNodeGroup());
+
+    cache.setSnapshot(SingleNodeGroup.GROUP, MULTIPLE_RESOURCES_SNAPSHOT2);
+
+    // Set a watch for the current snapshot with the same version but with resource hints present
+    // in the snapshot that the watch creator does not currently know about.
+    Map<String, Watch> watches = Resources.TYPE_URLS.stream()
+        .collect(Collectors.toMap(
+            typeUrl -> typeUrl,
+            typeUrl -> cache.createWatch(
+                ADS,
+                DiscoveryRequest.newBuilder()
+                    .setNode(Node.getDefaultInstance())
+                    .setTypeUrl(typeUrl)
+                    .setVersionInfo(MULTIPLE_RESOURCES_SNAPSHOT2.version(typeUrl))
+                    .addAllResourceNames(NAMES_MULTIPLE_RESOURCES_SNAPSHOT.get(typeUrl))
+                    .build(),
+                SNAPSHOT2.resources(typeUrl).keySet())));
+
+    // The snapshot version matches for all resources, but for eds and cds there are new resources present
+    // for the same version, so we expect the watches to trigger.
+    assertThatWatchReceivesSnapshot(watches.remove(Resources.CLUSTER_TYPE_URL),
+        MULTIPLE_RESOURCES_SNAPSHOT2);
+    assertThatWatchReceivesSnapshot(watches.remove(Resources.ENDPOINT_TYPE_URL),
+        MULTIPLE_RESOURCES_SNAPSHOT2);
+
+    // Remaining watches should not trigger
+    for (Watch watch : watches.values()) {
+      assertThatWatchIsOpenWithNoPendingResponses(watch);
+    }
+  }
+
+  @Test
+  public void successfullyWatchAllResourceTypesWithSetBeforeWatchWithSameRequestnVersionNewResourceHintsNoChange() {
+    SimpleCache<String> cache = new SimpleCache<>(new SingleNodeGroup());
+
+    cache.setSnapshot(SingleNodeGroup.GROUP, SNAPSHOT2);
+
+    // Set a watch for the current snapshot for the same version but with new resource hints not
+    // present in the snapshot that the watch creator does not know about.
+    Map<String, Watch> watches = Resources.TYPE_URLS.stream()
+        .collect(Collectors.toMap(
+            typeUrl -> typeUrl,
+            typeUrl -> cache.createWatch(
+                ADS,
+                DiscoveryRequest.newBuilder()
+                    .setNode(Node.getDefaultInstance())
+                    .setTypeUrl(typeUrl)
+                    .setVersionInfo(SNAPSHOT2.version(typeUrl))
+                    .addAllResourceNames(NAMES_MULTIPLE_RESOURCES_SNAPSHOT.get(typeUrl))
```

Comment:
Similar to above, I think this test would be a little more intuitive with an explicit comment here mentioning that the request is adding resource hints that are not part of `SNAPSHOT2`.